src/bin/scribble.ts

```diff
@@ -651,327 +703,319 @@ if ("version" in options) {
+    mergedUnits.forEach((sourceUnit) => {
+        if (contentsMap.has(sourceUnit.absolutePath)) {
+            rewriteImports(sourceUnit, contentsMap, factory);
+        }
+    });
 
-            for (const annots of annotMap.values()) {
-                for (const annot of annots) {
-                    filesWithAnnots.add(annot.originalSourceFile);
-                    nAnnots++;
-                }
-            }
+    /**
+     * Next try to instrument the merged SourceUnits.
+     */
+    const instrCtx = new InstrumentationContext(
+        factory,
+        mergedUnits,
+        assertionMode,
+        options["cov-assertions"],
+        addAssert,
+        callgraph,
+        cha,
+        filterOptions,
+        dedup(flatten(annotMap.values())),
+        new Map(),
+        contentsMap,
+        compilerVersionUsed,
+        debugEvents,
+        new Map(),
+        outputMode,
+        typeEnv,
+        semMap,
+        interposingQueue
+    );
 
-            if (nAnnots === 0) {
-                console.log(`Found ${nAnnots} annotations.`);
-            } else {
-                console.log(
-                    `Found ${nAnnots} annotations in ${filesWithAnnots.size} different files.`
-                );
-            }
+    /**
+     * Check if there is an instrumentation already in-place
+     */
+    if (isMetaDataFile) {
+        const { instrumentation } = getFilesInMetaData(metaDataFile);
 
-            for (const warning of findDeprecatedAnnotations(
-                mergedUnits,
-                contentsMap,
-                compilerVersionUsed
-            )) {
-                console.error(ppWarning(warning).join("\n"));
+        for (const fileName of instrumentation) {
+            if (isInstrumented(fileName, instrumentationMarker)) {
+                error(`File "${fileName}" is already instrumented`);
             }
         }
-
+    } else {
         /**
-         * Walk over the computed CHA and compute:
-         *  1. The set of contracts that have contract invariants (as the map contractInvs)
-         *  2. The set of contracts that NEED contract instrumentation (because they, a parent of theirs, or a child of theirs has contract invariants)
+         * @todo Decide if this is enough instead of checking instr files.
          */
-        const contractsNeedingInstr = computeContractsNeedingInstr(cha, annotMap);
-        const factory = new ScribbleFactory(mergedCtx);
-
-        // Next we re-write the imports to fix broken alias references (Some
-        // Solidity versions have broken references imports).
-        mergedUnits.forEach((sourceUnit) => {
-            if (contentsMap.has(sourceUnit.absolutePath)) {
-                rewriteImports(sourceUnit, contentsMap, factory);
+        for (const unit of instrCtx.units) {
+            if (isInstrumented(unit.sourceEntryKey, instrumentationMarker)) {
+                error(`File "${unit.sourceEntryKey}" is already instrumented`);
```

Comment:
This seems more efficient then `if` branch above. @cd1m0 WDYT?

Reply:
lgtn